File: leetcode/C++/0046-m-permutations.cpp
// DFS
// Time: O(n! * n)
// Space: O(n!)
class Solution {
public:
    void helper(vector<int>& nums, vector<int>& path, vector<vector<int>>& res) {
        if (path.size() == nums.size()) {
            res.push_back(path);
            return;
        }
        for (int n : nums) {
            vector<int>::iterator it = find(path.begin(), path.end(), n);
            if (it != path.end())
                continue;
            path.push_back(n);
            helper(nums, path, res);
            path.pop_back();
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0)
            return res;
        vector<int> path;
        helper(nums, path, res);
        return res;
    }
};

// DFS + Swap
class Solution {
public: 
    void helper(vector<int>& nums, int pos, vector<vector<int>>& res) {
        if (pos == nums.size()) {
            res.push_back(nums);
            return;
        }
        for (int i = pos; i < nums.size(); i++) {
            swap(nums[i], nums[pos]);
            helper(nums, pos + 1, res);
            swap(nums[i], nums[pos]);
        }
    }
    
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        if (nums.size() == 0)
            return res;
        helper(nums, 0, res);
        return res;
    }
};

File: leetcode/Python/0046-m-permutations.py
"""
DFS
Time: O(n! * n)
Space: O(n!)
"""
class Solution_1:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        res = []
        self.helper(nums, [], res)
        return res

    def helper(self, nums, path, res):
        if len(path) == len(nums):
            res.append(path[:])
            return

        for n in nums:
            if n in path:
                continue
            path.append(n)
            self.helper(nums, path, res)
            path.pop()

"""
DFS + Swap
"""
class Solution_2:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        res = []
        self.helper(nums, 0, res)
        return res

    def helper(self, nums, pos, res):
        # base case:
        if pos == len(nums):
            res.append(nums[:])
            return

        # main cases:
        for i in range(pos, len(nums)):
            # swap: fix a position for going down
            nums[pos], nums[i] = nums[i], nums[pos]
            self.helper(nums, pos + 1, res)
            # swap back: free a position for going right
            nums[pos], nums[i] = nums[i], nums[pos]
